File: rust/azure_iot_operations_protocol/src/telemetry/cloud_event.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use std::{
    fmt::{self, Display, Formatter},
    str::FromStr,
};

use chrono::{DateTime, Utc};

const DEFAULT_CLOUD_EVENT_SPEC_VERSION: &str = "1.0";
const DEFAULT_CLOUD_EVENT_EVENT_TYPE: &str = "ms.aio.telemetry";

/// Enum representing the cloud event fields.
#[derive(Debug, Copy, Clone, PartialEq)]
pub enum CloudEventFields {
    /// Identifies the event. Producers MUST ensure that source + id is unique for each distinct event.
    /// If a duplicate event is re-sent (e.g. due to a network error) it MAY have the same id.
    /// Consumers MAY assume that Events with identical source and id are duplicates.
    Id,
    /// Identifies the context in which an event happened.
    /// Often this will include information such as the type of the event source,
    /// the organization publishing the event or the process that produced the event.
    /// The exact syntax and semantics behind the data encoded in the URI is defined by the event producer.
    Source,
    /// The version of the CloudEvents specification which the event uses. This enables the
    /// interpretation of the context. Compliant event producers MUST use a value of 1.0 when
    /// referring to this version of the specification.
    SpecVersion,
    /// Describes the type of event related to the originating occurrence.
    /// Often this attribute is used for routing, observability, policy enforcement, etc.
    /// The format of this is producer defined and might include information such as the version of the type
    EventType,
    /// Identifies the subject of the event in the context of the event producer (identified by source).
    /// In publish-subscribe scenarios, a subscriber will typically subscribe to events emitted by a source,
    /// but the source identifier alone might not be sufficient as a qualifier for any specific event if the source context has internal sub-structure.
    Subject,
    /// Timestamp of when the occurrence happened.
    /// If the time of the occurrence cannot be determined then this attribute MAY be set to some other time (such as the current time)
    /// by the `CloudEvents` producer,
    /// however all producers for the same source MUST be consistent in this respect.
    Time,
    ///  Content type of data value. This attribute enables data to carry any type of content,
    ///  whereby format and encoding might differ from that of the chosen event format.
    DataContentType,
    ///  Identifies the schema that data adheres to.
    ///  Incompatible changes to the schema SHOULD be reflected by a different URI.
    DataSchema,
}

impl Display for CloudEventFields {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            CloudEventFields::SpecVersion => write!(f, "specversion"),
            CloudEventFields::EventType => write!(f, "type"),
            CloudEventFields::Source => write!(f, "source"),
            CloudEventFields::Id => write!(f, "id"),
            CloudEventFields::Subject => write!(f, "subject"),
            CloudEventFields::Time => write!(f, "time"),
            CloudEventFields::DataContentType => write!(f, "datacontenttype"),
            CloudEventFields::DataSchema => write!(f, "dataschema"),
        }
    }
}

impl FromStr for CloudEventFields {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "id" => Ok(CloudEventFields::Id),
            "source" => Ok(CloudEventFields::Source),
            "specversion" => Ok(CloudEventFields::SpecVersion),
            "type" => Ok(CloudEventFields::EventType),
            "subject" => Ok(CloudEventFields::Subject),
            "dataschema" => Ok(CloudEventFields::DataSchema),
            "datacontenttype" => Ok(CloudEventFields::DataContentType),
            "time" => Ok(CloudEventFields::Time),
            _ => Err(()),
        }
    }
}

/// Cloud Event struct
///
/// Implements the cloud event spec 1.0.
/// See [CloudEvents Spec](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md).
#[derive(Builder, Clone)]
#[builder(setter(into), build_fn(validate = "Self::validate"))]
pub struct CloudEvent {
    /// Identifies the event. Producers MUST ensure that source + id is unique for each distinct
    /// event. If a duplicate event is re-sent (e.g. due to a network error) it MAY have the same
    /// id. Consumers MAY assume that Events with identical source and id are duplicates.
    pub id: String,
    /// Identifies the context in which an event happened. Often this will include information such
    /// as the type of the event source, the organization publishing the event or the process that
    /// produced the event. The exact syntax and semantics behind the data encoded in the URI is
    /// defined by the event producer.
    pub source: String,
    /// The version of the cloud events specification which the event uses. This enables the
    /// interpretation of the context. Compliant event producers MUST use a value of 1.0 when
    /// referring to this version of the specification.
    #[builder(default = "DEFAULT_CLOUD_EVENT_SPEC_VERSION.to_string()")]
    pub spec_version: String,
    /// Contains a value describing the type of event related to the originating occurrence. Often
    /// this attribute is used for routing, observability, policy enforcement, etc. The format of
    /// this is producer defined and might include information such as the version of the type.
    #[builder(default = "DEFAULT_CLOUD_EVENT_EVENT_TYPE.to_string()")]
    pub event_type: String,
    /// Identifies the subject of the event in the context of the event producer (identified by
    /// source). In publish-subscribe scenarios, a subscriber will typically subscribe to events
    /// emitted by a source, but the source identifier alone might not be sufficient as a qualifier
    /// for any specific event if the source context has internal sub-structure.
    #[builder(default = "None")]
    pub subject: Option<String>,
    /// Identifies the schema that data adheres to. Incompatible changes to the schema SHOULD be
    /// reflected by a different URI.
    #[builder(default = "None")]
    pub data_schema: Option<String>,
    /// Content type of data value. This attribute enables data to carry any type of content,
    /// whereby format and encoding might differ from that of the chosen event format.
    #[builder(default = "None")]
    pub data_content_type: Option<String>,
    /// Timestamp of when the occurrence happened. If the time of the occurrence cannot be
    /// determined then this attribute MAY be set to some other time (such as the current time) by
    /// the cloud event producer, however all producers for the same source MUST be consistent in
    /// this respect. In other words, either they all use the actual time of the occurrence or they
    /// all use the same algorithm to determine the value used.
    #[builder(default = "None")]
    #[builder(setter(skip))]
    pub time: Option<DateTime<Utc>>, // This is optional per spec, but we will always add it
}

impl CloudEventBuilder {
    fn validate(&self) -> Result<(), String> {
        let mut spec_version = DEFAULT_CLOUD_EVENT_SPEC_VERSION.to_string();

        if let Some(sv) = &self.spec_version {
            spec_version = sv.to_string();
        }
        // Future versions of the spec may have different requirements
        if spec_version == "1.0" {
            // Any missing required fields are caught by the builder
            if let Some(id) = &self.id {
                if id.is_empty() {
                    return Err("id cannot be empty".to_string());
                }
            }
            if let Some(source) = &self.source {
                if source.is_empty() {
                    return Err("source cannot be empty".to_string());
                }
            }
            if let Some(event_type) = &self.event_type {
                if event_type.is_empty() {
                    return Err("event_type cannot be empty".to_string());
                }
            }
        } else {
            return Err("Invalid spec_version".to_string());
        }
        Ok(())
    }
}

impl CloudEvent {
    /// Get Cloud Event as headers for MQTT message
    #[must_use]
    pub fn to_headers(self) -> Vec<(String, String)> {
        let mut headers = vec![
            (CloudEventFields::Id.to_string(), self.id),
            (CloudEventFields::Source.to_string(), self.source),
            (CloudEventFields::SpecVersion.to_string(), self.spec_version),
            (CloudEventFields::EventType.to_string(), self.event_type),
        ];
        if let Some(subject) = self.subject {
            headers.push((CloudEventFields::Subject.to_string(), subject));
        }
        if let Some(data_schema) = self.data_schema {
            headers.push((CloudEventFields::DataSchema.to_string(), data_schema));
        }
        if let Some(data_content_type) = self.data_content_type {
            headers.push((
                CloudEventFields::DataContentType.to_string(),
                data_content_type,
            ));
        }
        if let Some(time) = self.time {
            headers.push((CloudEventFields::Time.to_string(), time.to_rfc3339()));
        }
        headers
    }
}

impl Display for CloudEvent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "id: {} \n\
            source: {} \n\
            event_type: {} \n\
            subject: {:?} \n\
            data_schema: {:?} \n\
            data_content_type: {:?} \n\
            time: {:?}",
            self.id,
            self.source,
            self.event_type,
            self.subject,
            self.data_schema,
            self.data_content_type,
            self.time
        )
    }
}

#[cfg(test)]
mod tests {
    use test_case::test_case;

    use super::*;

    #[test_case(CloudEventFields::SpecVersion; "cloud_event_spec_version")]
    #[test_case(CloudEventFields::EventType; "cloud_event_type")]
    #[test_case(CloudEventFields::Source; "cloud_event_source")]
    #[test_case(CloudEventFields::Id; "cloud_event_id")]
    #[test_case(CloudEventFields::Subject; "cloud_event_subject")]
    #[test_case(CloudEventFields::Time; "cloud_event_time")]
    #[test_case(CloudEventFields::DataContentType; "cloud_event_data_content_type")]
    #[test_case(CloudEventFields::DataSchema; "cloud_event_data_schema")]
    fn test_cloud_event_to_from_string(prop: CloudEventFields) {
        assert_eq!(prop, CloudEventFields::from_str(&prop.to_string()).unwrap());
    }
}

File: dotnet/src/Azure.Iot.Operations.Protocol/Telemetry/CloudEvent.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using Azure.Iot.Operations.Protocol.Models;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace Azure.Iot.Operations.Protocol.Telemetry
{
    /// <summary>
    /// Implements the CloudEvent spec 1.0. The required fields are source, type, id and specversion.
    /// Id is required but we want to update it in the same instance. 
    /// See <a href="https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md">CloudEvent Spec</a>
    /// </summary>
    /// <param name="source"><see cref="Source"/></param>
    /// <param name="type"><see cref="Type"/></param>
    /// <param name="specversion"><see cref="SpecVersion"/></param>
    public class CloudEvent(Uri source, string type = "ms.aio.telemetry", string specversion = "1.0")
    {
        private const string CloudEventIdProperty = "id";
        private const string CloudEventSourceProperty = "source";
        private const string CloudEventSpecVersionProperty = "specversion";
        private const string CloudEventTypeProperty = "type";
        private const string CloudEventTimeProperty = "time";
        private const string CloudEventDataContentTypeProperty = "datacontenttype";
        private const string CloudEventSubjectProperty = "subject";
        private const string CloudEventDataSchemaProperty = "dataschema";

        /// <summary>
        /// Identifies the context in which an event happened.
        /// Often this will include information such as the type of the event source, 
        /// the organization publishing the event or the process that produced the event. 
        /// The exact syntax and semantics behind the data encoded in the URI is defined by the event producer.
        /// </summary>
        public Uri Source => source;

        /// <summary>The version of the CloudEvents specification which the event uses. 
        /// This enables the interpretation of the context. 
        /// Compliant event producers MUST use a value of 1.0 when referring to this version of the specification.
        /// </summary>
        public string SpecVersion => specversion;

        /// <summary>
        /// Contains a value describing the type of event related to the originating occurrence. 
        /// Often this attribute is used for routing, observability, policy enforcement, etc. 
        /// The format of this is producer defined and might include information such as the version of the type
        /// </summary>
        public string Type => type;

        /// <summary>
        ///  Identifies the event. Producers MUST ensure that source + id is unique for each distinct event. 
        ///  If a duplicate event is re-sent (e.g. due to a network error) it MAY have the same id. 
        ///  Consumers MAY assume that Events with identical source and id are duplicates.
        /// </summary>
        public string? Id { get; set; } = null!; // although id is required, we want update it in the same instance from the sender.

        /// <summary>
        /// Timestamp of when the occurrence happened. 
        /// If the time of the occurrence cannot be determined then this attribute MAY be set to some other time (such as the current time) 
        /// by the CloudEvents producer, 
        /// however all producers for the same source MUST be consistent in this respect. 
        /// </summary>
        public DateTime? Time { get; set; }

        /// <summary>
        ///  Content type of data value. This attribute enables data to carry any type of content, 
        ///  whereby format and encoding might differ from that of the chosen event format.
        /// </summary>
        public string? DataContentType { get; internal set; }

        /// <summary>
        /// Identifies the subject of the event in the context of the event producer (identified by source). 
        /// In publish-subscribe scenarios, a subscriber will typically subscribe to events emitted by a source, 
        /// but the source identifier alone might not be sufficient as a qualifier for any specific event if the source context has internal sub-structure.
        /// </summary>
        public string? Subject { get; set; }

        /// <summary>
        ///  Identifies the schema that data adheres to. 
        ///  Incompatible changes to the schema SHOULD be reflected by a different URI.
        /// </summary>
        public string? DataSchema { get; set; }

        /// <summary>
        /// Creates an MQTT message context containing user properties and content type from this CloudEvent.
        /// This method provides default values for id (new GUID), time (current UTC time), and specversion (1.0) if not already set.
        /// </summary>
        /// <returns>A <see cref="CloudEventMqttContext"/> containing the MQTT user properties and content type.</returns>
        /// <exception cref="ArgumentException">Thrown if required fields are missing or invalid.</exception>
        public CloudEventMqttContext CreateMqttMessageContext()
        {
            // Validate required fields
            if (Source == null)
            {
                throw new ArgumentException("CloudEvent Source is required and cannot be null.");
            }

            if (string.IsNullOrEmpty(Type))
            {
                throw new ArgumentException("CloudEvent Type is required and cannot be null or empty.");
            }

            if (string.IsNullOrEmpty(SpecVersion))
            {
                throw new ArgumentException("CloudEvent SpecVersion is required and cannot be null or empty.");
            }

            // Create user properties list
            List<MqttUserProperty> userProperties = new();

            // Add required fields
            userProperties.Add(new MqttUserProperty(CloudEventSpecVersionProperty, SpecVersion));
            userProperties.Add(new MqttUserProperty(CloudEventTypeProperty, Type));
            userProperties.Add(new MqttUserProperty(CloudEventSourceProperty, Source.ToString()));

            // Add id with default if not set
            string idValue = Id ?? Guid.NewGuid().ToString();
            userProperties.Add(new MqttUserProperty(CloudEventIdProperty, idValue));

            // Add time with default if not set
            DateTime timeValue = Time ?? DateTime.UtcNow;
            userProperties.Add(new MqttUserProperty(CloudEventTimeProperty, timeValue.ToString("yyyy-MM-ddTHH:mm:ssK", CultureInfo.InvariantCulture)));

            // Add optional fields if present
            if (!string.IsNullOrEmpty(Subject))
            {
                userProperties.Add(new MqttUserProperty(CloudEventSubjectProperty, Subject));
            }

            if (!string.IsNullOrEmpty(DataSchema))
            {
                userProperties.Add(new MqttUserProperty(CloudEventDataSchemaProperty, DataSchema));
            }

            // Note: DataContentType is not added to user properties per CloudEvents MQTT spec
            // It should be set as the MQTT message's ContentType property

            return new CloudEventMqttContext
            {
                MqttUserProperties = userProperties,
                MqttMessageContentType = DataContentType
            };
        }

        /// <summary>
        /// Creates a CloudEvent from MQTT message context (user properties and content type).
        /// </summary>
        /// <param name="mqttMessageContext">The MQTT message context containing user properties and content type.</param>
        /// <returns>A <see cref="CloudEvent"/> parsed from the MQTT message context.</returns>
        /// <exception cref="ArgumentNullException">Thrown if mqttMessageContext is null.</exception>
        /// <exception cref="ArgumentException">Thrown if required CloudEvent fields are missing or invalid.</exception>
        public static CloudEvent CreateFromMqttUserProperties(CloudEventMqttContext mqttMessageContext)
        {
            ArgumentNullException.ThrowIfNull(mqttMessageContext);

            if (mqttMessageContext.MqttUserProperties == null)
            {
                throw new ArgumentException("MqttUserProperties cannot be null.", nameof(mqttMessageContext));
            }

            // Helper function to safely get user property
            string? GetUserProperty(string propertyName)
            {
                return mqttMessageContext.MqttUserProperties
                    .FirstOrDefault(p => p.Name.Equals(propertyName, StringComparison.OrdinalIgnoreCase))?.Value;
            }

            // Validate and get required fields
            string? specVersion = GetUserProperty(CloudEventSpecVersionProperty);
            if (string.IsNullOrEmpty(specVersion))
            {
                throw new ArgumentException($"CloudEvent '{CloudEventSpecVersionProperty}' is required.");
            }

            if (!specVersion.Equals("1.0", StringComparison.Ordinal))
            {
                throw new ArgumentException($"Only CloudEvent spec version 1.0 is supported. Version provided: {specVersion}");
            }

            string? id = GetUserProperty(CloudEventIdProperty);
            if (string.IsNullOrEmpty(id))
            {
                throw new ArgumentException($"CloudEvent '{CloudEventIdProperty}' is required.");
            }

            string? sourceValue = GetUserProperty(CloudEventSourceProperty);
            if (string.IsNullOrEmpty(sourceValue))
            {
                throw new ArgumentException($"CloudEvent '{CloudEventSourceProperty}' is required.");
            }

            if (!Uri.TryCreate(sourceValue, UriKind.RelativeOrAbsolute, out Uri? sourceUri))
            {
                throw new ArgumentException($"CloudEvent '{CloudEventSourceProperty}' must be a valid URI. Value provided: {sourceValue}");
            }

            string? typeValue = GetUserProperty(CloudEventTypeProperty);
            if (string.IsNullOrEmpty(typeValue))
            {
                throw new ArgumentException($"CloudEvent '{CloudEventTypeProperty}' is required.");
            }

            // Create the CloudEvent with required fields
            CloudEvent cloudEvent = new(sourceUri, typeValue, specVersion)
            {
                Id = id
            };

            // Parse optional time field
            string? timeValue = GetUserProperty(CloudEventTimeProperty);
            if (!string.IsNullOrEmpty(timeValue))
            {
                if (!DateTime.TryParse(timeValue, CultureInfo.InvariantCulture, DateTimeStyles.AdjustToUniversal, out DateTime parsedTime))
                {
                    throw new ArgumentException($"CloudEvent '{CloudEventTimeProperty}' must be a valid RFC3339 date-time. Value provided: {timeValue}");
                }
                cloudEvent.Time = parsedTime;
            }

            // Parse optional fields
            cloudEvent.Subject = GetUserProperty(CloudEventSubjectProperty);
            cloudEvent.DataSchema = GetUserProperty(CloudEventDataSchemaProperty);
            cloudEvent.DataContentType = mqttMessageContext.MqttMessageContentType;

            return cloudEvent;
        }
    }

    /// <summary>
    /// Represents the MQTT message context for a CloudEvent, including user properties and content type.
    /// </summary>
    public class CloudEventMqttContext
    {
        /// <summary>
        /// Gets or sets the MQTT user properties that represent CloudEvent fields.
        /// </summary>
        public List<MqttUserProperty> MqttUserProperties { get; set; } = new();

        /// <summary>
        /// Gets or sets the MQTT message content type, which corresponds to the CloudEvent datacontenttype field.
        /// </summary>
        public string? MqttMessageContentType { get; set; }
    }
}
